option-bucket/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+    ///
+    /// let mut x = Some(2);
+    /// let mut some = SomeBucket::new(&mut x).unwrap();
+    ///
+    /// let old = some.replace(5);
+    /// assert_eq!(old, 2);
+    /// assert_eq!(x, Some(5));
+    /// ```
+    pub fn replace(&mut self, value: T) -> T {
+        core::mem::replace(self.as_mut(), value)
+    }
+
+    /// Takes the value out of the option, leaving a `None` in its place,
+    /// and consuming this `SomeBucket`.
+    ///
+    /// ```
+    /// # use option_bucket::SomeBucket;
+    ///
+    /// let mut x = Some(vec![1, 2]);
+    /// let some = SomeBucket::new(&mut x).unwrap();
+    ///
+    /// let y = some.take();
+    /// assert_eq!(x, None);
+    /// assert_eq!(y, vec![1, 2]);
+    /// ```
+    pub fn take(self) -> T {
+        // SAFETY: `outer` is guaranteed to be `Some`
+        // by the invariants of `new_unchecked`
+        unsafe { self.outer.take().unwrap_unchecked() }
+    }
+}
+
+/// Abstraction for an [`&mut Option`][Option] that's known to be `None`.
+///
+/// # Size
+///
+/// `NoneBucket` is the same size as `&mut Option`
+pub struct NoneBucket<'a, T> {
+    outer: &'a mut Option<T>,
+}
+impl<'a, T> NoneBucket<'a, T> {
+    /// Creates a new [`NoneBucket`], without checking that
+    /// the input [`Option`] is `None`.
+    ///
+    /// It's recommended to use [`NoneBucket::new`] or
+    /// [`OptionBucket::new`] instead.
+    ///
+    /// # Safety
+    ///
+    /// Caller must guarantee that `opt` is NOT [`Some`].
+    pub unsafe fn new_unchecked(opt: &'a mut Option<T>) -> Self {
+        debug_assert!(
+            opt.is_none(),
+            "Undefined Behavior: `Some` value passed to `NoneBucket::new_unchecked`."
+        );
+
+        NoneBucket { outer: opt }
+    }
+
+    /// Creates a new [`NoneBucket`]. Returns `Some(NoneBucket<T>)`
+    /// if `opt` is [`None`], otherwise returns `None`.
+    ///
+    /// For an unchecked version, see [`NoneBucket::new_unchecked`].
+    pub fn new(opt: &'a mut Option<T>) -> Option<Self> {
+        if opt.is_none() {
+            // SAFETY: if conditional ensures that `opt` is `None`
+            unsafe { Some(Self::new_unchecked(opt)) }
+        } else {
+            None
+        }
+    }
+
+    /// Inserts value into the option, then returns a mutable reference to it.
+    ///
+    /// This is practically identical to [`Option::insert`], but avoids
+    /// operations handling [`drop`]ping the old value
+    /// (since we know there was no old value).
+    /// ```
+    /// # use option_bucket::NoneBucket;
+    ///
+    /// let mut opt = None;
+    /// let mut none = NoneBucket::new(&mut opt).unwrap();
+    /// let val = none.insert(1);
+    /// assert_eq!(*val, 1);
+    /// *val = 3;
+    /// assert_eq!(opt.unwrap(), 3);
+    /// ```
+    pub fn insert(self, value: T) -> &'a mut T {
+        // SAFETY: `outer` is `None`, so there is no old value to `drop`
+        unsafe {
+            let outer: *mut Option<T> = self.outer;
```

Comment:
Hm, in terms of invariants, I don't believe there is any hazard with performing a regular overwrite here and avoid the extra `unsafe`. Any perf uplift should also be negligible since 1) the compiler might be able to statically determine that it's `None` and 2) checking the `None` discriminator is fairly cheap all things considering.

It might also be nice to test it like this:
```
let old = self.outer.replace(Some(value));
debug_assert!(old.is_none(), "oops");
```

Reply:
> Hm, in terms of invariants, I don't believe there is any hazard with performing a regular overwrite here and avoid the extra `unsafe` (unless I'm missing something).

Certainly, there's no hazard in treating a `None` as an `Option`.

> Any perf uplift avoiding the drop should also be negligible since 1) the compiler might be able to statically determine that it's `None` and 2) checking the `None` discriminator is fairly cheap all things considering.

While that is true, I also don't really see value in removing the unsafe usage, given all of the unsafe usage in `SomeBucket`. If it was the difference between some unsafe and zero unsafe, that would be a different story.

While it may seem insignificant, the entry API exists specifically to allow the programmer to avoid hitting these kinds of unnecessary checks. I'd like to keep it the way it is.